compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -313,12 +315,122 @@ getTypeClassDictionary world = \case
     _ ->
         Nothing
 
-simplifyExpr :: World -> Expr -> Expr
-simplifyExpr world = fst . cata go
+-- | Go inside a closed expression and try to inline and beta reduce
+-- types and arguments, to try monomorphizing.
+inlineClosedExpr :: Int -> World -> Expr -> Simplifier Expr
+inlineClosedExpr d world e = do
+    e' <- go e
+    if e == e'
+        then pure e
+        else simplifyExpr' (d-1) e'
+  where
+    go = \case
+        e@(EVal x) -> do
+            case lookupValue x world of
+                Left _ -> pure e
+                Right dval -> pure (dvalBody dval)
+
+        ETmApp e1 e2 -> do
+            e1' <- go e1
+            case stripLoc e1' of
+                ETmLam (x,_) b -> do
+                    pure (applySubstInExpr (exprSubst x e2) b)
+                _ ->
+                    pure (ETmApp e1' e2)
+
+        ETyApp e t -> do
+            e' <- go e
+            case stripLoc e' of
+                ETyLam (x,_) b -> do
+                    pure (applySubstInExpr (typeSubst x t) b)
+                _ ->
+                    pure (ETyApp e' t)
+
+        ELocation l e -> ELocation l <$> go e
+
+        e -> pure e
+
+stripLoc :: Expr -> Expr
+stripLoc = \case
+    ELocation _ e -> stripLoc e
+    e -> e
+
+calcPartyLiterals :: Expr -> HasNoPartyLiterals
+calcPartyLiterals e = HasNoPartyLiterals (cata go e)
   where
-
-    go :: ExprF (Expr, Info) -> (Expr, Info)
     go = \case
+        EBuiltinF (BEParty _) -> False
+        f -> and f
+
+-- | Attempt to lift a closed expression to the top level. Returns either
+-- a variable expression that references the lifted expression, or
+-- returns the original expression.
+liftClosedExpr :: Int -> Expr -> Simplifier Expr
+liftClosedExpr d e = do
+    cache <- gets sCache
+    case Map.lookup e cache of
+        Just name -> do
+            EVal <$> selfQualify name
+
+        Nothing -> do
+            world <- gets sWorldExtended
+            version <- gets sVersion
+            case runGamma world version (typeOf' e) of
+                Right ty -> do
+                    name <- freshExprVarNameFor e
+                    e' <- inlineClosedExpr d world e
+                    addDefValue DefValue
+                        { dvalBinder = (name, ty)
+                        , dvalBody = e'
+                        , dvalLocation = Nothing
+                        , dvalNoPartyLiterals = calcPartyLiterals e'
+                        , dvalIsTest = IsTest False
+                        }
+                    EVal <$> selfQualify name
+
+                Left _ ->
```

Comment:
Can this ever happen? If not, should we throw an error here?

Reply:
It can happen when the information in the world is incomplete (e.g. it's part of a strongly connected component of definitions, or we are building with --incremental=yes). I can add a comment.